### PR TITLE
Add tracking signals for getting request/response bodies.

### PR DESCRIPTION
**Pull request by kowalski**
This is PR for issue: https://github.com/aio-libs/aiohttp/issues/2755

### CODE REVIEW

File: aiohttp/http_writer.py [hunk 1]
```diff
@@ -56,13 +59,18 @@ def _write(self, chunk):
...
 
-    def write(self, chunk, *, drain=True, LIMIT=64*1024):
+    def write(self, chunk, *, drain=True, LIMIT=64 * 1024):
         """Writes chunk of data to a stream.
 
         write_eof() indicates end of stream.
         writer can't be used after write_eof() method being called.
         write() return drain future.
         """
+        self.on_chunk_sent.freeze()
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: I pretty sure there is better place for signal freezing than first write attempt.

File: aiohttp/http_writer.py [hunk 2]
```diff
@@ -56,13 +59,18 @@ def _write(self, chunk):
             raise asyncio.CancelledError('Cannot write to closing transport')
         self._transport.write(chunk)
 
-    def write(self, chunk, *, drain=True, LIMIT=64*1024):
+    def write(self, chunk, *, drain=True, LIMIT=64 * 1024):
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: Don't touch if you don't change the logic.
If you want update code style -- do it in a separate PR.
For this particular line I pretty happy with status quo.

File: aiohttp/http_writer.py [hunk 3]
```diff
@@ -56,13 +59,18 @@ def _write(self, chunk):
...
         """Writes chunk of data to a stream.
 
         write_eof() indicates end of stream.
         writer can't be used after write_eof() method being called.
         write() return drain future.
         """
+        self.on_chunk_sent.freeze()
+        self.loop.create_task(
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: Spawning a new task is not what we need.
Technically `.write` is a coroutine, we need convert it to genuine async function.
Maybe I'll do it very soon in separate PR.

File: aiohttp/client_reqrep.py [hunk 4]
```diff
@@ -573,6 +587,9 @@ def __init__(self, method, url, *,
         self._auto_decompress = auto_decompress
         self._cache = {}  # reqired for @reify method decorator
 
+        # avoid circular reference so that __del__ works
+        self.on_chunk_received = Signal(owner=None)
```
Comment (asvetlov) [inline aiohttp/client_reqrep.py]: Not sure if we need a signal here.
It looks like signal concept overusage: you are creating `on_chunk_received` signal with the only usage by aiohttp internals itself. It is not very effective, signal subscription is literally a waste of time.
I suggest passing `traces` directly to response object.

File: aiohttp/http_writer.py [hunk 5]
```diff
@@ -31,6 +32,8 @@ def __init__(self, protocol, transport, loop):
         self._compress = None
         self._drain_waiter = None
 
+        self.on_chunk_sent = Signal(self)
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: Again pass traces instead of creating a new internal-only signal.

**Comment by asvetlov** [comment]
Documentation should be updated as well

File: aiohttp/http_writer.py (hunk 2 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: That's wasn't me, it's my emacs ;) I will revert this change

File: aiohttp/http_writer.py (hunk 5 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: Well, please consider that `StreamWriter` is used both by the http client and server. If I pass the `traces` in `__init__()` I will have to make something like:
```
def write( ):
    for trace in self.traces:
        await trace.sent_request_chunk_sent(...)
    ...
```
Now if you consider that code in the context of usage by http server, it clearly stops making sense....
This is why I introduced a signal and names it `on_chunk_sent`. Without specifying if it's http request it's sending or http response (as a server).

File: aiohttp/client_reqrep.py (hunk 4 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: Here the approach with passing `traces` will actually work quite well. I've introduced the signal to make it symetrical to the situation in request. I will update this to pass `traces` and remove the signal

File: aiohttp/http_writer.py (hunk 3 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: Are you sure we don't want this in another task ? Whatever handler does asynchronously will slow down writing the response. 

Also `.write` isn't actually coroutine, because it's not called like:
```
                for chunk in self.body:
                    writer.write(chunk)
```
(without `await`) The couroutine returned is never triggered.
I guess I would have to track all the places there `StreamWriter.write` is used and add `await` in there. That's not undoable, but it's rather big refactor.

File: aiohttp/http_writer.py (hunk 1 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: Actually I was quite surprised with the whole concept of `freeze()`. It's very tricky! Do you know why it's necessary at all ? 
The only reason I can think of is that the author wanted to prevent modifications to list of handlers while processing `send()`. 

Still, it's possible that we will decide to remove this `Signal` altogether, but if we don't, what do you think is the correct place to call freeze() ? 
- I can't do this in `__init__` because I want to add handlers later
- I can't delegate this responsibility to the user of the class, because this introduces a braking change where a code which works now like:
  ```
  writer = StreamWriter(...)
  await writer.write(...)
  ```
  will start raising `RuntimeError`

File: aiohttp/http_writer.py (hunk 5 above)
Comment (asvetlov) [reply aiohttp/http_writer.py]: Let's avoid signals for internal-only usage.
Here we have two options: pass a callback for sending signals or derive a `ClientStreamWriter` class. Personally I prefer the second approach.

File: aiohttp/http_writer.py (hunk 3 above)
Comment (asvetlov) [reply aiohttp/http_writer.py]: Yes, I'm sure. Spawning a task without waiting for result has an ugly smell.
More important spawning a task may reorder received signals, e.g. user may get *request chunk* after *request finished*.

`.write` should be async method. I'm working on this refactoring (required for #2698 anyway).
I suggest waiting for my job done.

File: aiohttp/http_writer.py (hunk 5 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: I will remove the `Signal` in favor of accepting an optional callable to be used as a callback. 
I prefer not to create a child class for this - too many layers of inheritance hurts code readability.

File: aiohttp/http_writer.py (hunk 3 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: Got it, I will wait until the refactoring is merged.

File: aiohttp/http_writer.py (hunk 2 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: Fixed

Comment (kowalski) [reply aiohttp/http_writer.py]: Also I fixed `setup.cfg` so that this line is not altered by anyone else who uses `autopep8`

File: aiohttp/http_writer.py (hunk 5 above)
Comment (asvetlov) [reply aiohttp/http_writer.py]: Ok

File: aiohttp/http_writer.py (hunk 2 above)
Comment (asvetlov) [reply aiohttp/http_writer.py]: Cool

**Comment by asvetlov** [comment]
@kowalski #2774 is done, `write()` is async now

**Comment by kowalski** [comment]
@asvetlov all fixes done, including adding documentation. 
I'm not squashing commits so that you can see what changes. I can squash changes after review  to keep history clean.

File: aiohttp/client_reqrep.py [hunk 6]
```diff
@@ -475,7 +477,14 @@ def keep_alive(self):
             if self.url.raw_query_string:
                 path += '?' + self.url.raw_query_string
 
-        writer = StreamWriter(conn.protocol, conn.transport, self.loop)
+        async def on_chunk_sent(chunk):
```
Comment (asvetlov) [inline aiohttp/client_reqrep.py]: Convert the function into ClientRequest's method. No need to create a nested function on every HTTP request.

File: aiohttp/http_writer.py [hunk 7]
```diff
@@ -62,6 +70,9 @@ def _write(self, chunk):
         writer can't be used after write_eof() method being called.
         write() return drain future.
         """
+        if self._on_chunk_sent:
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: Please do check `self._on_chunk_sent is not None`. It is more idiomatic.

File: aiohttp/tracing.py [hunk 8]
```diff
@@ -9,13 +9,14 @@
...
 
 __all__ = (
-    'TraceConfig', 'TraceRequestStartParams', 'TraceRequestEndParams',
-    'TraceRequestExceptionParams', 'TraceConnectionQueuedStartParams',
-    'TraceConnectionQueuedEndParams', 'TraceConnectionCreateStartParams',
-    'TraceConnectionCreateEndParams', 'TraceConnectionReuseconnParams',
-    'TraceDnsResolveHostStartParams', 'TraceDnsResolveHostEndParams',
-    'TraceDnsCacheHitParams', 'TraceDnsCacheMissParams',
-    'TraceRequestRedirectParams'
+    'TraceConfig', 'TraceRequestStartParams',
```
Comment (asvetlov) [inline aiohttp/tracing.py]: Do you really need to touch all these lines?
Appending two classes to the end of sequence should be enough.

File: docs/tracing_reference.rst [hunk 9]
```diff
@@ -147,6 +147,20 @@ TraceConfig
 
       ``params`` is :class:`aiohttp.TraceRequestStartParams` instance.
 
+   .. attribute:: on_request_chunk_sent
```
Comment (asvetlov) [inline docs/tracing_reference.rst:160]: Please also upgrade flow diagrams at the beginning of file.

**Comment by asvetlov** [comment]
Not squashing is fine, I'll do it on merging anyway. Well, github button will do actually.

File: aiohttp/client_reqrep.py [hunk 10]
```diff
@@ -168,7 +168,8 @@ def __init__(self, method, url, *,
                  proxy=None, proxy_auth=None,
                  timer=None, session=None, auto_decompress=True,
                  ssl=None,
-                 proxy_headers=None):
+                 proxy_headers=None,
+                 traces=[]):
```
Comment (kxepal) [inline aiohttp/client_reqrep.py]: Let's not use mutables as defaults. These may cause some nasty bugs we could easily avoid.

File: aiohttp/client_reqrep.py [hunk 11]
```diff
@@ -572,6 +583,7 @@ def __init__(self, method, url, *,
         self._timer = timer if timer is not None else TimerNoop()
         self._auto_decompress = auto_decompress
         self._cache = {}  # reqired for @reify method decorator
+        self._traces = traces
```
Comment (kxepal) [inline aiohttp/client_reqrep.py]: Why in ClientRequest traces is public attribute while in ClientResponse it's private?

File: aiohttp/client_reqrep.py (hunk 6 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: done

File: aiohttp/client_reqrep.py [hunk 12]
```diff
@@ -555,7 +565,8 @@ class ClientResponse(HeadersMixin):
 
     def __init__(self, method, url, *,
                  writer=None, continue100=None, timer=None,
-                 request_info=None, auto_decompress=True):
+                 request_info=None, auto_decompress=True,
+                 traces=[]):
```
Comment (kxepal) [inline aiohttp/client_reqrep.py]: Same defaults story.

File: aiohttp/http_writer.py (hunk 7 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: changed

File: aiohttp/tracing.py (hunk 8 above)
Comment (kowalski) [reply aiohttp/tracing.py]: Ok, reverted and added new classes at the end

File: docs/tracing_reference.rst (hunk 9 above)
Comment (kowalski) [reply docs/tracing_reference.rst:160]: done. Here I added new signals to description. Please let me know if you meant something more elaborate.
zrzut ekranu z 2018-02-27 15-04-09

File: aiohttp/client_reqrep.py (hunk 11 above)
Comment (asvetlov) [reply aiohttp/client_reqrep.py]: I can answer: `ClientResponse` is a public class visible by user.
The attr should be private.
About request I don't care.

Comment (kxepal) [reply aiohttp/client_reqrep.py]: Ok, fine for me.

File: docs/tracing_reference.rst [hunk 13]
```diff
@@ -34,8 +34,8 @@ Overview
      exception[shape=flowchart.terminator, description="on_request_exception"];
 
      acquire_connection[description="Connection acquiring"];
-     got_response;
-     send_request;
+     got_response[description="on_response_chunk_received"];
```
Comment (asvetlov) [inline docs/tracing_reference.rst]: Please rename `got_response` and `send_request` nodes to be close to actual signal names. Maybe that long names could be shortened on diagram -- be creative.

Also I like to see arrows what explicitly points that chunk_sent and chunk_received events can come more than once.

File: aiohttp/client_reqrep.py (hunk 10 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: true, changed to `None`

File: aiohttp/client_reqrep.py (hunk 11 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: Agreed. I switched now to `_traces` in both classes

File: aiohttp/client_reqrep.py (hunk 12 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: also changed to `None`

**Comment by asvetlov** [comment]
Wait, send_request and got_response in my mind is mostly for headers. We need different nodes for body chunks.

**Comment by asvetlov** [comment]
Please add `..versionadded:: 3.1` for new signals and data classes

File: setup.cfg [hunk 14]
```diff
@@ -1,5 +1,6 @@
 [pep8]
 max-line-length=79
+ignore=E225,E226
```
Comment (kxepal) [inline setup.cfg]: I'm in doubt that we should ignore these. Any reasons why to disable them globally?

Comment (kowalski) [reply setup.cfg]: Well, I'm not in place to judge this. I did it to be able to save a file which has `64*1000` literal without spaces around `*`. Otherwise autopep just fixes it for me.

**Comment by kowalski** [comment]
@asvetlov 
If I split nodes for headers and chunks this is what I get. What do you think ? 
zrzut ekranu z 2018-02-27 15-25-33

File: setup.cfg (hunk 14 above)
Comment (kxepal) [reply setup.cfg]: That autopep suggestion looks correct. You may also write `64000` instead without any loss in readability.

Comment (kowalski) [reply setup.cfg]: I'm with you.  @asvetlov specifically asked me to keep it as was.

Comment (asvetlov) [reply setup.cfg]: For me the strange thing is successful passing flake8 checks without the `ignore` setting.
I'm totally fine with 64000 or even better 0x10000. Pretty sure it should be 64KiB instead of 64kB

**Comment by asvetlov** [comment]
The diagram looks better but I don't see arrows from sent/recevied nodes to exception.
Maybe resulting picture will be unreadable, in this case I suggest extracting headers/body signals into separate diagram -- like I did for connection establishment and DNS lookup.
I don't know -- please do the best but if you'll fail I'll polish the diagram after merging.

Also the diagram shows that we need `on_request_headers_sent` and `on_response_headers_received` signals maybe :) -- but in separate PR obviously.

File: docs/tracing_reference.rst [hunk 15]
```diff
@@ -147,6 +155,23 @@ TraceConfig
...
+      ``params`` is :class:`aiohttp.TraceRequestChunkSentParams` instance.
+
+   .. attribute:: on_response_chunk_received
+
+
+      Property that gives access to the signals that will be executed
+      when a chunk of response body is received.
+
+      ``params`` is :class:`aiohttp.TraceResponseChunkReceivedParams` instance.
+
```
[the remark this answers is not in the document]
Comment (kowalski) [reply docs/tracing_reference.rst:177]: done

**Comment by kowalski** [comment]
@asvetlov done

File: aiohttp/client_reqrep.py [hunk 16]
```diff
@@ -475,7 +477,10 @@ def keep_alive(self):
             if self.url.raw_query_string:
                 path += '?' + self.url.raw_query_string
 
-        writer = StreamWriter(conn.protocol, conn.transport, self.loop)
+        writer = StreamWriter(
+            conn.protocol, conn.transport, self.loop,
+            on_chunk_sent=self._on_chunk_request_sent
```
[the remark this answers is not in the document]
Comment (kowalski) [reply aiohttp/client_reqrep.py:484]: `StreamWriter` is used in (at least) two scenarios:
- here, for writing request body
- in RequestHandler for writing response to incoming request

In the later case, having `traces` an calling `on_request_chunk_sent` seems wrong, because the class is used for writting something completely different. So `on_chunk_sent` is introduced to have something generic.

[the remark this answers is not in the document]
Comment (pfreixes) [reply aiohttp/client_reqrep.py:484]: Oks good point!

File: aiohttp/client_reqrep.py [hunk 17]
```diff
@@ -209,6 +210,7 @@ def __init__(self, method, url, *,
         if data or self.method not in self.GET_METHODS:
             self.update_transfer_encoding()
         self.update_expect_continue(expect100)
+        self._traces = traces or []
```
Comment (asvetlov) [inline aiohttp/client_reqrep.py]: In other aiohttp code we usually use another pattern:
```
if traces is None:
    traces = []
self._traces = traces
```
Please change the one-liner to explicit check.
It is a very subtle optimization for if `traces` is an empty list already you code calculates `False` for `bool(traces)` check and creates an empty list again.

File: aiohttp/client_reqrep.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/client_reqrep.py:484]: +1

File: aiohttp/client_reqrep.py [hunk 18]
```diff
@@ -572,6 +583,7 @@ def __init__(self, method, url, *,
         self._timer = timer if timer is not None else TimerNoop()
         self._auto_decompress = auto_decompress
         self._cache = {}  # reqired for @reify method decorator
+        self._traces = traces or []
```
Comment (asvetlov) [inline aiohttp/client_reqrep.py]: The same: replace `or` with explicit check for `is None` please

File: aiohttp/http_writer.py [hunk 19]
```diff
@@ -16,7 +17,12 @@
 
 class StreamWriter(AbstractStreamWriter):
 
-    def __init__(self, protocol, transport, loop):
+    def __init__(self, protocol, transport, loop, on_chunk_sent=None):
+        assert (
```
Comment (asvetlov) [inline aiohttp/http_writer.py]: The assert is in the hot-pass route.
I understand your striving for type checking but all `StreamWriter` usage is under the library control, we always should pass the proper values for `on_chunk_sent` parameter.
Please remove the assert statement.

File: docs/tracing_reference.rst [hunk 20]
```diff
@@ -34,16 +34,26 @@ Overview
      exception[shape=flowchart.terminator, description="on_request_exception"];
 
      acquire_connection[description="Connection acquiring"];
-     got_response;
-     send_request;
+     headers_received;
```
Comment (asvetlov) [inline docs/tracing_reference.rst:37]: The diagram is cryptic a little  but I can live with it.

File: setup.cfg (hunk 14 above)
Comment (asvetlov) [reply setup.cfg]: Please drop the change. Feel free to replace the limit with `0x10000` if needed -- I'm +-0 for it.

**Comment by asvetlov** [comment]
@kxepal @pfreixes please review again

**Comment by kxepal** [comment]
I agree with already requested changes, so nothing more could add.

**Comment by kxepal** [comment]
@asvetlov 
Ok, approving to not be a blocker. It seems you're on pulse of further changes here.

**Comment by asvetlov** [comment]
@kxepal thanks

**Comment by pfreixes** [comment]
Just missing the proper tests in `test_http_write.py` and `test_client_response.py` that must check the changes at the function level, for the default path - `traces` and `on_chunk_sent` as None - and 
 the none default  path. We alredy have it for `test_connection.py` right now.

I know that globally new handlers are tested via `test_client_session.py`, but its more an integration test, local tests allow us to check al branches explicitly.

**Comment by asvetlov** [comment]
@pfreixes agree.
@kowalski please add required tests

File: setup.cfg (hunk 14 above)
Comment (kowalski) [reply setup.cfg]: done

File: aiohttp/client_reqrep.py (hunk 18 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: done

File: aiohttp/client_reqrep.py (hunk 17 above)
Comment (kowalski) [reply aiohttp/client_reqrep.py]: done

File: aiohttp/http_writer.py (hunk 19 above)
Comment (kowalski) [reply aiohttp/http_writer.py]: done

File: aiohttp/http_writer.py [hunk 21]
```diff
@@ -62,6 +70,9 @@ def _write(self, chunk):
         writer can't be used after write_eof() method being called.
         write() return drain future.
         """
+        if self._on_chunk_sent is not None:
+            await self._on_chunk_sent(chunk)
+
```
[the remark this answers is not in the document]
Comment (kowalski) [reply aiohttp/http_writer.py:69]: done

File: aiohttp/client_reqrep.py [hunk 22]
```diff
@@ -796,6 +808,8 @@ def _notify_content(self):
         if self._content is None:
             try:
                 self._content = await self.content.read()
+                for trace in self._traces:
+                    await trace.send_response_chunk_received(self._content)
```
[the remark this answers is not in the document]
Comment (kowalski) [reply aiohttp/client_reqrep.py:816]: done

**Comment by kowalski** [comment]
all done

**Comment by pfreixes** [comment]
Thanks !

**Comment by kowalski** [comment]
@asvetlov done (added `./CHANGES/2767.feature`)

**Comment by asvetlov** [comment]
Great job, thanks to everyone.
Will merge when tests pass.
